[TensorPipe] Use registry for transports and channels

lw opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#39957 [TensorPipe] Use registry for transports and channels**

In order to provide a pluggable and extendable way to add new transports and channels to the TensorPipe agent, we use two registries. This allows us to separate the specific details of each backend (e.g., how it determines what address to use) from the generic logic of setting up TensorPipe.

Differential Revision: [D22017614](https://our.internmc.facebook.com/intern/diff/D22017614/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D22017614/)!

lw committed:
[TensorPipe] Use registry for transports and channels

In order to provide a pluggable and extendable way to add new transports and channels to the TensorPipe agent, we use two registries. This allows us to separate the specific details of each backend (e.g., how it determines what address to use) from the generic logic of setting up TensorPipe.

Differential Revision: [D22017614](https://our.internmc.facebook.com/intern/diff/D22017614/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D22017614/)!

[ghstack-poisoned]

mrshenli commented on torch/csrc/distributed/rpc/tensorpipe_agent.cpp:
```diff
@@ -27,7 +27,14 @@ const std::string kServerActiveAsyncCalls = "agent.server_active_async_calls";
-std::string guessAddress(tensorpipe::transport::uv::Context& uvContext) {
+} // namespace
+
+C10_DEFINE_REGISTRY(TensorPipeTransportRegistry, TransportRegistration);
```
What does `C10_DEFINE_REGISTRY`/`C10_DECLARE_REGISTRY ` Do? Shall we have some comments here?

lw replied:
They are defined in https://github.com/pytorch/pytorch/blob/master/c10/util/Registry.h and there's already some comments in that file... do you think there's any more info we could add here without repeating ourselves?

In a nutshell, a registry is a string-to-generator map. C10_DECLARE_REGISTRY adds the _declaration_ of a registry, i.e., the part that must go in headers. C10_DEFINE_REGISTRY adds the _definition_, i.e., the part that must go in source files.

mrshenli replied:
I see. If a 3rd-party user would like to register a new transport/channel, will they go through the same or a different process?

lw replied:
At the moment we're not prioritizing that specific use case, which means that, yes, for now to add a new transport/channel users need to write a file where they include `C10_REGISTER_CREATOR(TensorPipeChannelRegistry, their_name, ...)` and then link it together with PyTorch. It is indeed quite clunky. With c10d::Extension we would be able to provide a second path to register transport/channels, which would be more user-friendly.

lw replied:
Note that also the API of transports/channels isn't really finalized (we're not making explicit properties about its stability, it should be considered private), so users might incur other problems if they were to trying to plug in their custom channel.

mrshenli commented on torch/csrc/distributed/rpc/tensorpipe_agent.cpp:
```diff
@@ -50,6 +57,50 @@ std::string guessAddress(tensorpipe::transport::uv::Context& uvContext) {
+
+C10_REGISTER_CREATOR(TensorPipeChannelRegistry, basic, makeBasicChannel);
+
+#ifdef TP_ENABLE_CMA
```
Curious, what is a CMA channel?

lw replied:
CMA stands for cross-memory attach, it's a funny syscall that Linux has (and, as far as I know, no other OS) which allows one process to read from/write to another process's private memory (provided the right permissions are in place). In a sense, this is pretty much what GDB does when you attach it to a running process.

Let me add a quick comment explaining what each transport/channel does.

mrshenli commented on torch/csrc/distributed/rpc/tensorpipe_agent.cpp:
```diff
@@ -133,36 +184,40 @@ TensorPipeAgent::~TensorPipeAgent() {
-  context_->registerChannel(
-      0, "cma", std::make_shared<tensorpipe::channel::cma::Context>());
-#endif
+  for (auto& key : TensorPipeTransportRegistry()->Keys()) {
```
I assume the keys here will be tcp/shm/cma? What if I want to add two tcp channels for two different NICs?

lw replied:
Indeed, the keys at the moment are ones of the registry, which means that for now only one instance of each transport/channel is allowed, and thus we cannot have two TCP channels with different configurations. This is one of the things we need to figure out, probably by v1.7, related to the agent options. For example, we could make it so that the transports/channels in registry are "templates" that the user can instantiate multiple times if desired, like:
```
options=rpc.TensorPipeAgentOptions(transports={
    "tcp_nic0": Transport("uv", device="eth0"),
    "tcp_nic1": Transport("uv", device="eth1"),
})
```

mrshenli commented on torch/csrc/distributed/rpc/tensorpipe_agent.cpp:
```diff
@@ -133,36 +184,40 @@ TensorPipeAgent::~TensorPipeAgent() {
-  VLOG(1) << "RPC agent for " << workerInfo_.name_ << " is using address "
-          << uvAddress;
+  std::vector<std::string> addresses;
+  int lowestPriority = -1;
```
Would I be correct if I assume higher the number lower the priority?

lw replied:
Very good catch, thanks! This was indeed the case, because of a long-standing bug of TensorPipe where, counterintuitively, lower priorities took precedence over higher ones. I ended up fixing that bug just before sending out this diff (https://github.com/pytorch/tensorpipe/commit/8fb1fe66f8fe7d95f4b3aa934afc9b194cd47db6) but forgot to update this constant.

mrshenli commented on torch/csrc/distributed/rpc/tensorpipe_agent.cpp:
```diff
@@ -133,36 +184,40 @@ TensorPipeAgent::~TensorPipeAgent() {
+  for (auto& key : TensorPipeChannelRegistry()->Keys()) {
+    std::unique_ptr<ChannelRegistration> reg =
+        TensorPipeChannelRegistry()->Create(key);
+    context_->registerChannel(
```
Where there be any error if the Transport presents but channel is absent (or vice versa)?

lw replied:
Yes, there should be at least one transport and one channel defined. If that is not the case, then an error would be raised somewhere, but I suspect this is not currently handled very gracefully, so probably it would just end in an uncaught exception.

Note however that a transport is necessary already just to open a pipe: the client must specify an address to connect to, and that address is specific to the transport (for example, the address could be `uv://127.0.0.1:1234` or `shm://some_unix_domain_socket_name`). So for the user to be able to use one such address it must have registered `uv` or `shm` already.

As for channels, it is possible that registering no channels would work if none of the messages sent over the pipes contain tensors (i.e., they only contain their payload, and thus only operate on scalars). It's not something that is supported/recommended. Also, the `basic` channel should be safe to always keep installed, because all it does is just re-use the transport to act as a channel, and is thus a sort-of fallback.

mrshenli commented on torch/csrc/distributed/rpc/tensorpipe_agent.cpp:
```diff
@@ -133,36 +184,40 @@ TensorPipeAgent::~TensorPipeAgent() {
   const std::vector<uint8_t> selfAddrData(address.begin(), address.end());
   nameToAddressStore_.set(workerInfo_.name_, selfAddrData);
 
+  VLOG(1) << "RPC agent for " << workerInfo_.name_ << " is using address "
```
#39974 added `operator<<` for `WorkerInfo`, just in case you want to use it here.

lw replied:
Thanks for the pointer, it probably makes sense to use that. However we're doing this all over the file so it's probably best to change all these occurrences at once in a separate diff, rather than polluting this one here.

mrshenli commented on torch/csrc/distributed/rpc/tensorpipe_agent.cpp:
```diff
@@ -133,36 +184,40 @@ TensorPipeAgent::~TensorPipeAgent() {
-  const auto address = listener_->url("tcp");
+  const auto address = listener_->url(lowestPriorityTransport);
   const std::vector<uint8_t> selfAddrData(address.begin(), address.end());
   nameToAddressStore_.set(workerInfo_.name_, selfAddrData);
```
Curious about the future plan here. The current code looks like only registers the transport with the lowest priority to the name->address store. What about other transport? And what if this transport only connects this worker with a subset of the gang? How would the rest of the gang talk to this worker?

lw replied:
So, each pipe has an "initial" transport, which is the one specified in the address used to connect (e.g., "uv" in `uv://127.0.0.1:1234` or "shm" in `shm://foobar`), which is the transport used to bootstrap the pipe. Afterwards, during the handshake, the pipe will try to negotiate a better transport, if it can find one. "Better" means a transport with a priority higher than the initial transport. "If it can find one" means a transport that both sides can use and that would work (e.g., SHM would not work if the two sides are on different machines). If such a transport is found the pipe switches over to that one, otherwise it remains on the initial one. (Note that the addresses for these other transports do not have to be stored in the nameToAddressStore, because the server will send them to the client over the pipe as part of the handshake).

What this means is that any transport that has a _lower_ priority than the initial transport will never be used (it makes no sense to downgrade to it, it will just make things worse). Therefore, the "converse" property, is that the initial transport should be the transport with lowest priority.

It also means that the transport with lowest priority should be (the only) one that is able to connect all workers. If that were not the case, then it means that there is another transport with higher priority (and thus better!) that works for all workers, but then we should make that transport the default and remove all the lower-priority ones.

All the above reasoning is based on the assumption that there is a strict total order of "goodness" among all transports, which is captured by the priority. This assumption works, for now, but is not always correct. You could imagine having both a TCP and an InfiniBand transport: the TCP one might provide a higher bandwidth, while the InfiniBand one might provide a lower latency. So there isn't one that is strictly better than the other, and thus we cannot rank them using priorities. This kind of more fine-grained selection is something we want to work on soon.

mrshenli reviewed (approved):
Code change LGTM! I just have some questions inline.

lw committed:
Update on "[TensorPipe] Use registry for transports and channels"


In order to provide a pluggable and extendable way to add new transports and channels to the TensorPipe agent, we use two registries. This allows us to separate the specific details of each backend (e.g., how it determines what address to use) from the generic logic of setting up TensorPipe.

Differential Revision: [D22017614](https://our.internmc.facebook.com/intern/diff/D22017614/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D22017614/)!

[ghstack-poisoned]
